### PR TITLE
Make TaintIssuesSynchronizer branch-aware

### LINKED ISSUE #3410: Make taint branch-aware
### Description

Make fetching taint issues branch-aware

- import and call `IServerBranchProvider`
- fix up tests

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Fixes #3410

### CODE REVIEW

File: src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs [hunk 1]
```diff
@@ -52,6 +52,7 @@ public void MefCtor_CheckIsExported()
                 MefTestHelpers.CreateExport<ISonarQubeService>(),
                 MefTestHelpers.CreateExport<ITaintIssueToIssueVisualizationConverter>(),
                 MefTestHelpers.CreateExport<IConfigurationProvider>(),
+                MefTestHelpers.CreateExport<IServerBranchProvider>(),
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs:55]: New required import for TaintIssuesSynchronizer

File: src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs [hunk 2]
```diff
@@ -118,27 +119,31 @@ public void SynchronizeWithServer_DisconnectedInTheMiddle_ServerInfoIsReusedAndN
         }
 
         [TestMethod]
-        public async Task SynchronizeWithServer_NotInConnectedMode_StoreCleared()
+        public async Task SynchronizeWithServer_NotInConnectedMode_ServerComponentsNotCalledAndStoreCleared()
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs:122]: Renamed to make it clearer what the test is checking.
It didn't seem worth having a separate test for each server component.

File: src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs [hunk 3]
```diff
@@ -118,27 +119,31 @@ public void SynchronizeWithServer_DisconnectedInTheMiddle_ServerInfoIsReusedAndN
...
+                serverBranchProvider: serverBranchProvider.Object,
                 logger: logger);
 
             await testSubject.SynchronizeWithServer();
 
             CheckStoreIsCleared(taintStore);
 
+            // Server components should not be called
             logger.AssertPartialOutputStringExists("not in connected mode");
-            sonarQubeServer.Invocations.Count.Should().Be(0);
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs:140]: Cosmetic: switched existing checks to use `HaveCount`, as this is what is suggested by the Fluent Assertions analyzers.

File: src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs [hunk 4]
```diff
@@ -312,6 +317,26 @@ public async Task SynchronizeWithServer_StandaloneMode_UIContextIsCleared()
             toolWindowServiceMock.Invocations.Count.Should().Be(0);
         }
 
+        [TestMethod]
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs:320]: One new test to check that the taint synchronizer fetches and passes on the expected branch name.

File: src/IssueViz.Security/Taint/TaintIssuesSynchronizer.cs [hunk 5]
```diff
@@ -92,8 +95,10 @@ public async Task SynchronizeWithServer()
                 }
 
                 var projectKey = bindingConfiguration.Project.ProjectKey;
+                var serverBranch = await serverBranchProvider.GetServerBranchNameAsync();
+
                 var taintVulnerabilities = await sonarQubeService.GetTaintVulnerabilitiesAsync(projectKey,
-                    null, // TODO - pass the current branch
+                    serverBranch,
```
Comment (duncanp-sonar) [inline src/IssueViz.Security/Taint/TaintIssuesSynchronizer.cs:101]: The actual product changes at this level are very simple.
Hopefully suppressions will be as straightforward...

**Review by ugras-ergun-sonarsource — APPROVED** [review]

**Pull request merged**